crates/rattler_installs_packages/src/python_env/byte_code_compiler.rs

```diff
@@ -0,0 +1,279 @@
+            .as_ref()
+            .expect("the channel is only dropped on drop")
+            .send(source_path.to_owned())
+            .map_err(|_| CompilationError::HostQuit)
+    }
+
+    /// Compile the specified python file and wait for the compilation to finish.
+    pub fn compile_and_wait(&self, source_path: &Path) -> Result<PathBuf, CompilationError> {
+        let (tx, rx) = channel::<CompilationResponse>();
+        self.compile(source_path, move |result| {
+            tx.send(result)
+                .expect("the reader is waiting on this response");
+        })?;
+        rx.recv()
+            .expect("the sender was dropped before sending a response")
+    }
+
+    /// Wait for all queued compilations to finish.
+    pub fn wait(mut self) -> Result<(), std::io::Error> {
+        // Drop the request channel to signal the compilation host that we are done. This will
+        // ensure that the stdin pipe of the compilation host is closed which will signal the host
+        // to exit.
+        drop(self.request_tx.take());
+
+        // Wait for the compilation host to exit
+        self.child
+            .take()
+            .expect("the child is only dropped on drop")
+            .wait()?;
+
+        Ok(())
+    }
+}
+
+impl Drop for ByteCodeCompiler {
+    fn drop(&mut self) {
+        drop(self.request_tx.take());
+        if let Some(mut child) = self.child.take() {
+            child.wait().unwrap();
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::python_env::system_python_executable;
+
+    #[test]
+    fn test() {
+        let python_path = system_python_executable().unwrap();
+
+        // Create a temporary file that holds the sourcecode of the compilation host
+        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
+        compiler_source
+            .write_all(include_bytes!("compile_pyc.py"))
+            .unwrap();
+
+        // Spawn a compiler and compile the compilation host source code.
+        let compiler = ByteCodeCompiler::new(&python_path).unwrap();
+        let pyc_file = compiler.compile_and_wait(compiler_source.path()).unwrap();
+
+        // Make sure the compiled file exists
+        assert!(pyc_file.is_file(), "The compiled file does not exist!");
+
+        // Wait should immediately return
+        compiler.wait().unwrap();
+    }
+
+    #[test]
+    fn test_failed_case() {
+        let python_path = system_python_executable().unwrap();
+
+        // Create a temporary file that holds the sourcecode of the compilation host
+        let mut compiler_source = tempfile::Builder::new().suffix(".py").tempfile().unwrap();
+        compiler_source.write_all(b"$").unwrap();
+
+        // Spawn a compiler and compile the compilation host source code.
+        let compiler = ByteCodeCompiler::new(&python_path).unwrap();
+        compiler
+            .compile_and_wait(compiler_source.path())
+            .unwrap_err();
```

Comment:
Does this mean the process will fail when it can't compile a file? I think usually we should ignore failures – it can happen that some Python file can only be compiled with a newer version, or has some template variables in it etc. so we should silently fail (or with a trace::warn or so).

Reply:
Yes! I will change the unpack behavior to just ignore these errors.